L5 CLE gym envt added

theDebugger811 opened the pull request:
L5Kit Gym Environment created for closed loop training

perone commented on l5kit/l5kit/evaluation/error_functions.py:
```diff
@@ -15,3 +16,16 @@ def l2_error(pred: torch.Tensor, gt: torch.Tensor) -> torch.Tensor:
+
+def closest_angle(angle_a: torch.Tensor, angle_b: torch.Tensor) -> torch.Tensor:
+    """ Finds the closest angle between angle_b - angle_a in radians.
+        Returns an angle between [-pi, pi]
```
Move this phrase to the `:return:` below.

perone commented on l5kit/l5kit/simulation/unroll.py:
```diff
@@ -76,6 +76,15 @@ def __init__(self, scene_id: int, sim_dataset: SimulationDataset,
         self.ego_ins_outs = ego_ins_outs[scene_id]
         self.agents_ins_outs = agents_ins_outs[scene_id]
 
+        # Edit
+        # # Required for Bokeh Visualizer
+        # self.tls_frames = self.simulated_dataset.dataset.tl_faces
+        # self.agents_th = self.simulated_dataset.cfg["raster_params"]["filter_agents_threshold"]
+
+        # # Remove Dataset attributes
+        # self.recorded_dataset = None
+        # self.simulated_dataset = None
+
```
remove commented code

perone commented on l5kit/l5kit/simulation/unroll.py:
```diff
@@ -122,7 +131,8 @@ def __init__(self, sim_cfg: SimulationConfig, dataset: EgoDataset,
         self.sim_cfg = sim_cfg
         if not sim_cfg.use_ego_gt and model_ego is None:
-            raise ValueError("ego model should not be None when simulating ego")
+            print("ego model should not be None when simulating ego")
```
Is this not an error anymore ?

perone reviewed (commented):
Thanks @theDebugger811, I added a few comments, let's sync on the meeting today.

theDebugger811 commented on l5kit/Pipfile:
```diff
@@ -13,8 +13,9 @@ sphinx = "*"
 pre-commit = "*"
-importlib-metadata = "*"  # required for python 3.8
 typing-extensions = "*"
+line-profiler = "*"
+importlib-metadata = "*"
```
I use line-profiler to time the code. Is it okay to remove in the final merge? I will keep a note.

perone replied:
Leave it only for your development env then, you can remove from here.

perone commented on l5kit/l5kit/environment/cle_metricset.py:
```diff
@@ -0,0 +1,120 @@
+    def get_results(self) -> Dict[str, Any]:
+        """Perform all required aggregations and returns a dictionary
+        composed by all results."""
+        pass
```
raise a `NotImplementedError` here.

perone commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,200 @@
+
+        frame_index = self.frame_index
+        next_frame_index = frame_index + 1
+        episode_over = next_frame_index == (len(self.sim_dataset) - self.future_num_frames)  # Note !!
```
What is note here ?

perone commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,200 @@
+        # We are using "info" to output simulated outputs
+        info = {}
+        if done:
+            info = {"info": self.get_simulated_outputs()}
```
Do we need this ? This will be copied through IPC so it might be very expensive.

perone commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,200 @@
+        return simulated_outputs
+
+    def render(self) -> None:
+        pass
```
I think it would be better to raise a `NotImplementError` error here, to be explicit about the support of this.

perone commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,200 @@
+    def render(self) -> None:
+        pass
+
+    def close(self) -> None:
```
When is this called ?

perone commented on l5kit/l5kit/environment/reward.py:
```diff
@@ -0,0 +1,91 @@
+from l5kit.environment.cle_metricset import L5BaseMetricSet, SimulationOutputGym
+
+
+class Reward(object):
```
Please create an abstraction here: you will have `Reward` (that can be an ABC) and then you will have something like a `L2ImitationReward` (this concrete implementation of the `Reward`). Imagine that can exist multiple rewards later.

perone commented on l5kit/l5kit/environment/reward.py:
```diff
@@ -0,0 +1,91 @@
+        yaw_error = self.yaw_weight * torch.abs(scene_metrics['yaw_error_ca'][frame_index + 1])
+
+        # clip reward
+        reward = max(-self.rew_clip_thresh, -dist_error.item())
```
Add an option to enable/disable it.

perone commented on l5kit/l5kit/environment/test_gym.py:
```diff
@@ -0,0 +1,27 @@
+import os
```
This file should go to the examples folder if it is an example, otherwise should be an integration test.

perone reviewed (commented):
Added some more comments, things we already aligned that are missing is the unit-testing and check the linting that is failing due to a line size I think. Thanks a lot @theDebugger811 !

theDebugger811 commented:
Hi @perone, Thanks for the detailed review. I have incorporated the changes. Also, I added SimulationOutputCLE base class (that gets passed to Close-loop evaluator). All the tests are passing locally but not here. Any idea?

perone commented on examples/RL/test_gym.py:
```diff
@@ -0,0 +1,29 @@
+import os
```
Rename this file from `test_gym.py` to `create_gym_example.py`, files prefixed with `test_` are often captured by testing systems like `pytest`, etc, as being unit-tests that should be executed during testing.

perone commented on examples/RL/test_gym.py:
```diff
@@ -0,0 +1,29 @@
+from l5kit.environment.register_l5_env import create_l5_env
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--data_path', type=str, default='/home/ubuntu/level5_data/',
```
These paths do not exist in users computers.

perone commented on examples/RL/test_gym.py:
```diff
@@ -0,0 +1,29 @@
+parser = argparse.ArgumentParser()
+parser.add_argument('--data_path', type=str, default='/home/ubuntu/level5_data/',
+                    help='Environment configuration file')
+parser.add_argument('--env_config_path', type=str, default='/home/ubuntu/src/l5kit/examples/RL/config.yaml',
```
These paths do not exist in users computers.

perone commented on examples/RL/test_gym.py:
```diff
@@ -0,0 +1,29 @@
+os.environ["L5KIT_DATA_FOLDER"] = args.data_path
+
+# create and register L5Kit gym env
+create_l5_env(args.env_config_path, args.eps_length, args.disable_cle, args.rew_clip)
```
Isn't this supposed to be called on a `__init__.py` of l5kit ? Please see [this reference](https://github.com/openai/gym-soccer/blob/master/gym_soccer/__init__.py).

perone replied:
Note that there are some parameters here that should be supported in the `gym.make("L5-CLE-v0", [args])` args, for instance, if the user wants to change the episode length after registration. Reference for the `make()` -> https://github.com/openai/gym/blob/0.10.10/gym/envs/registration.py#L77.

perone commented on examples/RL/test_gym.py:
```diff
@@ -0,0 +1,29 @@
+create_l5_env(args.env_config_path, args.eps_length, args.disable_cle, args.rew_clip)
+
+# call
+env = gym.make('L5-CLE-v0')
```
Can you add some calls here to show the environment ? Example: get one initial observation and print the shapes.

perone commented on l5kit/l5kit/environment/cle_metricset.py:
```diff
@@ -0,0 +1,117 @@
+        self.agents_th = simulated_dataset.cfg["raster_params"]["filter_agents_threshold"]
+
+
+class CLEMetricSet(ABC):
```
Rename this class to `CLEBaseMetricSet` as this is the base (abstract base class) class.

perone commented on l5kit/l5kit/environment/cle_metricset.py:
```diff
@@ -0,0 +1,117 @@
+        raise NotImplementedError
+
+
+class L5BaseMetricSet(CLEMetricSet):
```
Can remove the `Base` from the name here.

perone commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,185 @@
+
+
+class GymStepOutput(NamedTuple):
+    """ The output dict of gym env.step
```
```suggestion
    """The output dict of gym env.step
```

perone commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,185 @@
+    """ Custom Environment of L5 Kit that can be registered in OpenAI Gym. """
+
+    def __init__(self, env_config_path: str, sim_cfg: SimulationConfig, reward: Reward, cle: bool) -> None:
+        """
+        :param env_config_path: path to the L5Kit environment configuration file
+        :param simulation_cfg: configuration of the L5Kit closed loop simulator
+        :param reward: calculates the reward for the gym environment
+        :param cle: flag to enable close loop environment updates
+        """
```
This docstring should to together with the one above (Custom Environment of ...)

perone commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,185 @@
+
+
+class L5Env(gym.Env):
+    """ Custom Environment of L5 Kit that can be registered in OpenAI Gym. """
```
```suggestion
    """Custom Environment of L5 Kit that can be registered in OpenAI Gym. """
```

perone commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,185 @@
+
+        # Define action and observation space
+        # Continuous Action Space: gym.spaces.Box (X, Y, Yaw * number of future states)
+        # self.action_space = spaces.Box(low=-1000, high=1000, shape=(self.future_num_frames * 3, ))
```
Old commented code.

perone commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,185 @@
+
+        # Scene Sampling config
+        self.max_scene_id = cfg["gym_params"]["max_scene_id"]
+        if cfg["gym_params"]["overfit"]:  # Overfit
```
This comment isn't adding much here, can remove.

perone commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,185 @@
+
+        # EGO
+        if not self.sim_cfg.use_ego_gt:
+            action = rescale_action(action)
+            action = convert_to_dict(action, self.future_num_frames)
```
These two functions `rescale_action` and `convert_to_dict` are very particular of the environment, can you move these functions to be methods of the environment ? They can be private prefixed w/ `_`, as in `def _rescale_action(...)`.

perone commented on l5kit/l5kit/environment/reward.py:
```diff
@@ -0,0 +1,141 @@
+
+
+class RewardInput(NamedTuple):
+    """ The input tuple to calculate reward
```
```suggestion
    """The input tuple to calculate reward
```

Remove the space in the beginning of the docstrings. [Reference for example](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html#an-example-class-with-docstrings).

perone commented on l5kit/l5kit/environment/utils.py:
```diff
@@ -0,0 +1,87 @@
+from l5kit.rasterization import Rasterizer
+
+
+def rescale_action(action: np.ndarray, x_mu: float = 1.20, x_scale: float = 0.2,
+                   y_mu: float = 0.0, y_scale: float = 0.03, yaw_scale: float = 3.14) -> np.ndarray:
```
Missing documentation.

perone commented on l5kit/l5kit/evaluation/error_functions.py:
```diff
@@ -15,3 +16,16 @@ def l2_error(pred: torch.Tensor, gt: torch.Tensor) -> torch.Tensor:
+
+
+def closest_angle(angle_a: torch.Tensor, angle_b: torch.Tensor) -> torch.Tensor:
+    """ Finds the closest angle between angle_b - angle_a in radians.
```
```suggestion
    """Finds the closest angle between angle_b - angle_a in radians.
```

perone reviewed (commented):
Thanks @theDebugger811, added some more comments. After that we need to add the unit-testing.

theDebugger811 commented:
@szagoruyko  @lucabergamini , could you please review this L5Kit gym-compatible environment PR?
Thanks!

lucabergamini commented on l5kit/l5kit/cle/metric_set.py:
```diff
@@ -0,0 +1,86 @@
+        raise NotImplementedError
+
+
+class L5MetricSet(BaseMetricSet):
```
shouldn't this be doing something in the methods?

theDebugger811 replied:
I did not understand. Should I shift L5MetricSet class to the 'environment' folder?

lucabergamini replied:
what is the purpose of this class?

theDebugger811 replied:
Its similar in spirit to the driveformer metric_set. [Ref](https://github.com/lyft/avjadoo/blob/master/experiments/planning/driveformer_cle_metricset.py)
Its a metric set for L5Kit. 

We can have different metric sets inheriting from this. Currently, we have one with distance and yaw error. In future, we can have one with collision metrics as well.

lucabergamini replied:
got it.

It's a little bit weird because this class does nothing so we could also move everything in the super. But I guess we need this hierarchy for inheritance at some point

lucabergamini commented on l5kit/l5kit/simulation/unroll.py:
```diff
@@ -105,12 +105,33 @@ def build_trajectory_states(frames: np.ndarray) -> torch.Tensor:
         return trajectory_states
 
 
+class SimulationOutput(SimulationOutputCLE):
```
I think this will break some visualisation utils (e.g. `simulation_out_to_visualizer_scene`)

theDebugger811 replied:
The refactored SimulationOutput class has the same attributes as the original SimulationOutput class.  The tests are running fine. Would you suggest to check something regarding breaking visualization utils? 

Christian suggested to make SimulationOutput inherent from SimulationOutputCLE so that close-loop evaluator-specific components are kept separated from visualization components. Let me know if there is a better way to do this.

lucabergamini commented on l5kit/l5kit/simulation/unroll.py:
```diff
@@ -119,11 +140,12 @@ def __init__(self, sim_cfg: SimulationConfig, dataset: EgoDataset,
         :param model_ego: the model to be used for ego
         :param model_agents: the model to be used for agents
         :param keys_to_exclude: keys to exclude from input/output (e.g. huge blobs)
+        :param verify_model: flag to verify the ego/agent model when use_ego_gt/use_agent_gt is False
```
Why is this required?

theDebugger811 replied:
The ClosedLoopSimulator requires both the SimulationDataset and an ego_model when use_ego_gt = False. However, in gym environment, the SimulationDataset (part of environment) is independent of the ego_model (the policy). So, the ClosedLoopSimulator should not have an ego_model (it is not part of environment), and this will raise an error. 

The verify_model flag helps to check whether we want to verify the presence of ego_model (as in L5Kit) or not (for gym). Does this explanation help?

lucabergamini replied:
A bit. It seems we're introducing a lot of complexity though

theDebugger811 replied:
Without this flag, we might have to end up writing a class very similar to ClosedLoopSimulator, but without the ego and agent models as the model (policy) in L5Gym is decoupled from the simulator (the environment).

I believe the verification flag is a small added complexity that allows to re-use the entire ClosedLoopSimulator for L5 gym environment. What do you think?

lucabergamini replied:
should we have this more explicit? 
what about:
- defining an IntEnum with 2 modes (gym, l5kit)
- taking that as a param and perform the check based on that

what do you think?

theDebugger811 replied:
Thats certainly a better way. I will do this!

lucabergamini commented on l5kit/l5kit/tests/conftest.py:
```diff
@@ -36,6 +36,19 @@ def cfg() -> dict:
     return load_config_data("./l5kit/tests/artefacts/config.yaml")
 
 
+@pytest.fixture(scope="function")
```
does this need to be function? Isn't enough once per session?

lucabergamini reviewed (commented):
left some comments

theDebugger811 commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,260 @@
+        return self
+
+
+class SimulationOutputGym(SimulationOutputCLE):
```
SimulationOutputCLE: Is used for close loop evaluation.
SimulationOutputGym: Is used for visualization.

SimulationOutputCLE contains parameters that suffice for Close loop evaluation.  SimulationOutputGym contains additional parameters that are compatible for Bokeh Visualizer as well. Does this help explain?

lucabergamini replied:
yeah, but the name does not, there is no relation between `gym` and `bokeh`

lucabergamini replied:
better :)

theDebugger811 replied:
Thanks, I will then update the class name.

theDebugger811 commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,260 @@
+        :return: the observation of first frame of sampled scene index
+        """
+        # Sample a episode randomly
+        self.scene_index = random.randint(0, self.max_scene_id)
```
`init` helps to initialize the environment. `reset` is used to start a new episode (by sampling a new scene to be rolled out). The attributes defined in `reset` are specific to that particular rollout, hence not a part of `init`. Should I still define these attributes in `init`?

theDebugger811 commented on l5kit/l5kit/environment/reward.py:
```diff
@@ -0,0 +1,88 @@
+        :return: the reward is the combination of L2 error from groundtruth trajectory and (optionally) yaw error
+        """
+        scene_id = simulated_outputs[0].scene_id
+        self.metric_set.evaluate(simulated_outputs)
```
Yes, I wanted to make use of existing code as much as possible. The distance error metrics support the compute for entire scene only. Do you think it is better to add an extra method in the `DisplacementErrorMetric` class to support per frame evaluation? It wont save a lot of compute I believe.

lucabergamini replied:
I'm a little bit worried about this, for the following reason:
- we're passing a metricset to just use `displacement_error_l2` and `yaw_error_closest_angle`. However, we're not checking if those metrics are in the metricset;
- we're hiding how this reward is computed, as it can only be understood by reading the `get_reward` carefully.

A partial solution could be:
- assert what we need from the metricset in the init (I'm still not convinced it is worth passing the full metricset, maybe we could pass the two metrics)

regarding calling this over the whole scene:
- why not slicing it before passing it to the evaluate instead that afterwards?

theDebugger811 replied:
@perone what would you suggest?

theDebugger811 commented on l5kit/l5kit/environment/reward.py:
```diff
@@ -0,0 +1,88 @@
+
+        # clip reward
+        reward = float(-dist_error.item())
+        if self.enable_clip:
```
We just want to clip the distance error (in x, y) and not the yaw error (it is bounded).

lucabergamini replied:
ok, add it as a comment

theDebugger811 commented on l5kit/l5kit/environment/reward.py:
```diff
@@ -0,0 +1,88 @@
+        raise NotImplementedError
+
+
+class CLE_Reward(Reward):
```
It's not technically supervised right? There is not back-propagation of penalty (-reward). Removed the `_` from class name.

theDebugger811 commented on examples/RL/create_gym_example.py:
```diff
@@ -0,0 +1,12 @@
+import os
```
Yes, for now it's a file. It will be replaced with a notebook (TODO).

lucabergamini replied:
add the TODO in the code then

theDebugger811 commented:
Thanks a lot @lucabergamini for the review and suggestions. 

@perone , @szagoruyko  I have incorporated Luca's review and previous comments by perone. Could you have another pass on the latest commit? Thanks!

szagoruyko commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,323 @@
+        if self.reset_scene_id is not None:
+            self.scene_index = self.reset_scene_id
+        else:  # Sample a scene
+            self.scene_index = random.randint(0, self.max_scene_id)
```
you need to use `self.np_random.randint` here. `random.randint` does not have the seed set.

szagoruyko reviewed (commented):
started reviewing, going to continue tomorrow

theDebugger811 commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,323 @@
+    :param eps_length: the number of step to simulate per episode in the gym environment.
+    """
+
+    def __new__(cls, eps_length: int = 32, start_frame_idx: int = 0) -> 'SimulationConfigGym':
```
I read online: [Ref](https://stackoverflow.com/questions/42385916/inheriting-from-a-namedtuple-base-class/42386174#42386174). It says we need to override __new__ in case of namedTuples.

szagoruyko commented on l5kit/l5kit/environment/utils.py:
```diff
@@ -0,0 +1,116 @@
+            output_data[k] = np.expand_dims(v.cpu().numpy(), axis=0)
+        else:
+            print(k, v, type(v))
+            raise NotImplementedError
```
best to avoid prints in the code for proper error handling
```suggestion
            raise NotImplementedError(f"{type(v)} is not supported (field {k})")
```

szagoruyko commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,323 @@
+
+        # Reset Kinematic model
+        if self.use_kinematic:
+            init_kin_state = np.array([0.0, 0.0, 0.0, 0.1 * ego_input[0]['curr_speed']])
```
why multiply by 0.1?

theDebugger811 replied:
Because the step_time is 0.1. I have added a comment there.

szagoruyko reviewed (commented):
added few more comments

theDebugger811 commented on examples/RL/create_gym_example.py:
```diff
@@ -0,0 +1,19 @@
+
+# By setting the L5KIT_DATA_FOLDER variable, we can point the script
+# to the folder where the data lies.
+os.environ["L5KIT_DATA_FOLDER"] = os.environ["HOME"] + '/level5_data'
```
Hi, it seems I keep missing the point what to do here. Could you please suggest an alternative? Thanks!

perone replied:
Just remove this line and change the comment to tell that the env variable L5KIT_DATA_FOLDER should be filled with the correct path for the download dataset.

theDebugger811 commented on l5kit/l5kit/environment/__init__.py:
```diff
@@ -0,0 +1,9 @@
+gym.envs.register(
+    id='L5-CLE-v0',
+    entry_point="l5kit.environment.envs.l5_env:L5Env",
+    kwargs={'env_config_path': '../../../examples/RL/config.yaml'},
```
Could you suggest a way to handle this error? This will also solve the previous comment regarding "having a default for `env_config_path` to be the `config.yaml `

theDebugger811 commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,323 @@
+    :param eps_length: the number of step to simulate per episode in the gym environment.
+    """
+
+    def __new__(cls, eps_length: int = 32, start_frame_idx: int = 0) -> 'SimulationConfigGym':
```
Assigning `use_ego_gt: bool = False` before `__new__` leads to an error. I have defined the default arguments inside `__new__`

perone replied:
Sorry I wasn't very clear, after doing this you should remove the `__new__` as you will have already defined the params, right ?

theDebugger811 replied:
I want to inherit from` SimulationConfig` here as other parts of the code (eg. `ClosedLoopSimulator` require type `SimulationConfig`.

perone replied:
But my suggestion was to still keep inheriting from it right ? Not sure if I understood.

theDebugger811 replied:
On making the requested changes, I got the following error:
`TypeError: <lambda>() missing 5 required positional arguments: 'use_ego_gt', 'use_agents_gt', 'disable_new_agents', 'distance_th_far', and 'distance_th_close'`

I believe due to inheritance, SimulationConfig is called and it does not have default values set for the above entries.  What do you suggest?

perone replied:
But we can add defaults in the SimulationConfig right ?

theDebugger811 commented on l5kit/l5kit/environment/utils.py:
```diff
@@ -0,0 +1,116 @@
+    :return: the unnormalized action
+    """
+    if use_kinematic:
+        return ActionRescaleParams()
```
`ActionRescaleParams()` is a namedTuple that defines the rescale parameters for both kinematic and non-kinematic params. Do you think having separate tuples `KinematicActionRescaleParams()` and `NonKinematicActionRescaleParams()` will help?

perone replied:
What I meant was to keep the same named tuple, but just the way you create it should be changed. If the default params are linked to the kinematic params, then instead create a function to return the `ActionRescaleParams` with the kinematic params.

theDebugger811 commented on examples/RL/create_gym_example.py:
```diff
@@ -0,0 +1,19 @@
+env_config_path = "./config.yaml"
+
+# call
+env = gym.make('L5-CLE-v0', env_config_path=env_config_path)
```
The default `env_config_path` is `None` so that environment can be registered (as per our discussion today).

I personally prefer to explicitly point out the config file being used for this example.

perone replied:
That's fine, thanks !

theDebugger811 commented:
@szagoruyko @perone Thanks for your feedback. I have incorporated the changes. @perone could you have another pass? Thanks!

perone commented on examples/RL/create_gym_example.py:
```diff
@@ -0,0 +1,22 @@
+# follow the instructions (https://github.com/lyft/l5kit#download-the-datasets)
+# to correctly organise it.
+
+# Set the L5KIT_DATA_FOLDER path and uncomment the line below
```
This comment is not correct, it is either "set the environment variable" or "uncomment line below", not both. Setting the env variable and then uncommenting the line below will replace what was set.

theDebugger811 replied:
Okay, just keeping "Uncomment the line below"

perone commented on l5kit/l5kit/cle/metric_set.py:
```diff
@@ -0,0 +1,86 @@
+    """
+
+    def __init__(self, metric_prefix: str = "L5") -> None:
+        """Constructor method
+        """
```
```suggestion
```

perone commented on l5kit/l5kit/cle/metrics.py:
```diff
@@ -237,3 +237,42 @@ def compute(self, simulation_output: SimulationOutput) -> torch.Tensor:
+    """
+    metric_name = "yaw_error"
+
+    def __init__(self, error_function: error_functions.ErrorFunction) -> None:
```
Can set the `error_function` here to be the default as the closest angle.

theDebugger811 replied:
I am following the template of `DisplacementErrorMetric`. It doesn't have a default `error_function`. Can I keep as it is currently?

perone replied:
For displacement error it makes more sense to have different metrics to compute, but for the yaw the one we're using it is a more natural option, users might confuse that and end up other metrics, so it should be ok to deviate here in this case.

theDebugger811 replied:
Okay, should I still keep the `YawErrorCAMetric` then?

perone commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,333 @@
+    The speed at start_frame_idx is always 0 (not indicative of the true current speed).
+    We therefore simulate the episode from (start_frame_idx + 1, start_frame_idx + eps_length + 1)
+    """
+    use_ego_gt: bool = False
+    use_agents_gt: bool = True
+    disable_new_agents: bool = False
+    distance_th_far: float = 30.0
+    distance_th_close: float = 15.0
+    num_simulation_steps: int = 33
+    start_frame_idx: int = 0
```
Can you add docs here so Sphinx can capture and generate automatically ? The format here should be something like:

```
(...)
#: doc for the variable goes here
start_frame_idx: int = 0
```

Another alternative is to add it in the docstr above, feel free to check which one looks better after Sphinx renders it.

perone commented on l5kit/l5kit/environment/reward.py:
```diff
@@ -0,0 +1,119 @@
+        self.metric_set = metric_set if metric_set is not None else L2DisplacementYawMetricSet()
+
+        # Verify that error metrics necessary for reward calculation are present in the metric set
+        metric_names = [metric.metric_name for metric in self.metric_set.evaluation_plan.metrics]
```
There is already a `metrics_dict` in the `EvaluationPlan`, so you don't have to collect metric names here, just get the dict and then do a `metric_name not in dict` to check.

perone commented on l5kit/l5kit/environment/reward.py:
```diff
@@ -0,0 +1,119 @@
+        """
+        self.metric_set.reset()
+
+    def slice_simulated_output(self, index: int,
```
This looks like a static method, can you add the `staticmethod` decorator and remove the self ? As it is not using  `self` for anything.

perone commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,333 @@
+            frame_index = 0  # Dummy final obs (when episode_over)
+
+        ego_input = self.sim_dataset.rasterise_frame_batch(frame_index)
+        self.ego_input_dict = convert_to_numpy(ego_input[0])
```
Same comment here, the `rasterise_frame_batch` return type is `List[Dict[str, np.ndarray]]`, so it seems to be already a numpy type correct ?

perone commented on l5kit/l5kit/visualization/visualizer/zarr_utils.py:
```diff
@@ -224,12 +225,102 @@ def simulation_out_to_visualizer_scene(sim_out: SimulationOutput, mapAPI: MapAPI
+        if has_ego_info:
+            # print(frame_idx)
+            ego_in_out = ego_ins_outs[frame_idx]
+            # print("Target at: ", frame_idx, ego_in_out.inputs['target_positions'], ego_in_out.inputs['target_yaws'])
+            # print("Output at: ", frame_idx, ego_in_out.outputs['positions'], ego_in_out.outputs['yaws'])
```
Same here

perone reviewed (commented):
Thanks @theDebugger811, added a few minor more comments.

perone commented on l5kit/l5kit/environment/envs/l5_env.py:
```diff
@@ -0,0 +1,333 @@
+        # Output first observation
+        self.frame_index = 1  # Frame_index 1 has access to the true ego speed
+        ego_input = self.sim_dataset.rasterise_frame_batch(self.frame_index)
+        self.ego_input_dict = convert_to_numpy(ego_input[0])
```
I think collate might be misleading here as it is not doing the same as the torch collate does. I wouldn't add another function just to expand the dimensions, can be done inline in the code as needed as it is a simple op.

theDebugger811 commented:
@perone thanks for the quick review. I have incorporated the suggestions, removed the debug comments.
